### ISSUE TITLE
 parallel computing

### ISSUE THREAD

**Issue by zhaolongjun** [usnistgov/REFPROP-manager#4]
[test.txt](https://github.com/usnistgov/REFPROP-manager/files/3110345/test.txt)

In order to realize parallel computing，I loaded three DLLs, and the results of single-threaded running were correct and fast, but multithreaded running was time-consuming and error-prone. Then I tried to add thread locks, but the results were still incorrect and the time overhead increased. 

error： [TPFLSH error 215] supercritical or vapor density iteration did not converge:  [TPRHO error 203] vapor iteration has not converged for T =  732.50 K, P =  10.001 MPa, rho (last guess) =   2.3025 mol/L, x (mol frac) = 1.00000    
[PQFLSH error] (sat vapor):  [SATP error 144] iteration for saturation state did not converge; P =  10.100 MPa.# 2 #t 728.3 h 998
 
![image](https://user-images.githubusercontent.com/18252866/56629141-fabb4480-667e-11e9-90d9-3e04417d8561.png)

![image](https://user-images.githubusercontent.com/18252866/56629251-5b4a8180-667f-11e9-9922-70ec7670e3da.png)

**Comment by ianhbell** [comment]
Can you please copy-paste your test into the issue window on github?  The txt file seems to be using some funky encoding that my computer doesn’t understand.

This should definitely not be the behavior, I wonder what is going on.  Did you do the SETUP in serial?  The calls to SETUP reads/writes to files, and could result in funky behavior in theory, but because the REFPROP versions are in theory at least entirely independent, I can’t see why this would happen.  Did you try my example?

**Comment by zhaolongjun** [comment]
#include "REFPROP-manager/REFPROP-manager.hpp"
#include <stdio.h>
#include <pthread.h>
#include <unistd.h>
#include <time.h>

using namespace  std;

//创建互斥锁的全局变量
pthread_mutex_t myMutex1;
pthread_mutex_t myMutex2;
pthread_mutex_t myMutex3;

//初始化锁
int CreateMutexLock(pthread_mutex_t *mutex,pthread_mutexattr_t *attr)
{
    int iReturnValue=-1;
    //初始化成功返回0，否则返回非0
    iReturnValue=pthread_mutex_init(mutex,attr);
    return iReturnValue;
}
//加锁
int AddMutexLock(pthread_mutex_t *mutex)
{
    int iReturnValue=-1;
    iReturnValue=pthread_mutex_lock(mutex);
    return iReturnValue;
}
//解锁
int UnlockMutexLock(pthread_mutex_t *mutex)
{
    int iReturnValue=-1;
    iReturnValue=pthread_mutex_unlock(mutex);
    return iReturnValue;
}

//销毁锁
int DestoryMutexLock(pthread_mutex_t *mutex)
{
    int iReturnValue=-1;
    iReturnValue=pthread_mutex_destroy(mutex);
    return iReturnValue;
}

//测试函数1
void* TestFunc1(void*)
{
    //初始化锁
    if(0!=CreateMutexLock(&myMutex1,NULL))
    {
        cout<<"Create mutex lock fail.\n";
        return nullptr;
    }
    //加锁
    if(0!=AddMutexLock(&myMutex1))
    {
        cout<<"Add mutex lock fail.\n";
        return nullptr;
    }

    {
        int ierr = 0;
        char herr[255];
        int kq = 1;
        double z[] = { 1.0 }, x[] = { 1.0 }, y[] = { 1.0 }, T = 700, p = 10100.325, d = -1, dl = -1, dv = -1, h = -1, s = -1, u = -1, cp = -1, cv = -1, q = 1, w = -1;

        double timeStart = (double) clock();

        for(int i = 0; i < 1000; i++)
        {
            p += 0.001;
            PQFLSHdll(1,0, p, q, z, kq, T, d, dl, dv, x, y, u, h, s, cp, cv, w, ierr, herr, 255);
            cout<<"# 1 #"<<"p "<<p<< " h "<<h<<endl;
            if(ierr != 0)
            {
                cout<<"# 1 #"<<" ierr "<<ierr<<" herr "<<herr;
            }
        }

        double timeFinsh = (double) clock();
        cout<<"# 1 #"<<"Time = "<<timeFinsh- timeStart<<endl;
    }


    //解锁
    if(0!=UnlockMutexLock(&myMutex1))
    {
        cout<<"Unlock mutex lock fail.\n";
        return nullptr;
    }

    //销毁锁
    if(0!=DestoryMutexLock(&myMutex1))
    {
        cout<<"Destory mutex lock fail.\n";
        return nullptr;
    }
//    //线程自身调用pthread_exit()函数，终止线程
//    pthread_exit(0);
}
//测试函数2
void* TestFunc2(void*)
{
    //初始化锁
    if(0!=CreateMutexLock(&myMutex2,NULL))
    {
        cout<<"Create mutex lock fail.\n";
        return nullptr;
    }

    //加锁
    if(0!=AddMutexLock(&myMutex2))
    {
        cout<<"Add mutex lock fail.\n";
        return nullptr;
    }

    {
        int ierr = 0;
        char herr[255];
        double z[] = {1.0}, x[] = {1.0}, y[] = {1.0}, T= 700, p = 10001.325, d = -1, dl = -1, dv = -1, h = -1, s = -1, u = -1, cp = -1, cv = -1, q = -1, w = -1;

        double timeStart = (double) clock();

        for(int i = 0; i < 1000; i++)
        {
            T +=0.1;
            TPFLSHdll(2, 0, T, p, z, d, dl, dv, x, y, h,s,u,cp,cv,w,q,ierr,herr,255);
            cout<<"# 2 #"<<"t "<<T<< " h "<<h<<endl;
            if(ierr != 0)
            {
                cout<<"# 2 #"<<" ierr "<<ierr<<" herr "<<herr;
            }
        }

        double timeFinsh = (double) clock();
        cout<<"# 2 #"<<"Time = "<<timeFinsh- timeStart<<endl;
    }

    //解锁
    if(0!=UnlockMutexLock(&myMutex2))
    {
        cout<<"Unlock mutex lock fail.\n";
        return nullptr;
    }
    //销毁锁
    if(0!=DestoryMutexLock(&myMutex2))
    {
        cout<<"Destory mutex lock fail.\n";
        return nullptr;
    }
//    //线程自身调用pthread_exit()函数，终止线程
//    pthread_exit(0);
}

//测试函数3
void* TestFunc3(void*)
{
    //初始化锁
    if(0!=CreateMutexLock(&myMutex3,NULL))
    {
        cout<<("Create mutex lock fail.\n");
        return nullptr;
    }

    //加锁
    if(0!=AddMutexLock(&myMutex3))
    {
        cout<<("Add mutex lock fail.\n");
        return nullptr;
    }

    double timeStart = (double) clock();
    {
        {
            int ierr = 0;
            char herr[255];
            int kq = 1;
            double z[] = { 1.0 }, x[] = { 1.0 }, y[] = { 1.0 }, T = 700, p = 10100.325, d = -1, dl = -1, dv = -1, h = -1, s = -1, u = -1, cp = -1, cv = -1, q = 1, w = -1;
            for(int i = 0; i < 1000; i++)
            {
                p += 0.001;
                PQFLSHdll(3,0, p, q, z, kq, T, d, dl, dv, x, y, u, h, s, cp, cv, w, ierr, herr, 255);
                cout<<"# 3 #"<<"p "<<p<< " h "<<h<<endl;
                if(ierr != 0)
                {
                    cout<<"# 3 #"<<" ierr "<<ierr<<" herr "<<herr;
                }
            }
        }

        {
            int ierr = 0;
            char herr[255];
            double z[] = {1.0}, x[] = {1.0}, y[] = {1.0}, T= 700, p = 10001.325, d = -1, dl = -1, dv = -1, h = -1, s = -1, u = -1, cp = -1, cv = -1, q = -1, w = -1;
            for(int i = 0; i < 1000; i++)
            {
                T +=0.1;
                TPFLSHdll(3, 0, T, p, z, d, dl, dv, x, y, h,s,u,cp,cv,w,q,ierr,herr,255);
                cout<<"# 3 #"<<"t "<<T<< " h "<<h<<endl;
                if(ierr != 0)
                {
                    cout<<"# 3 #"<<" ierr "<<ierr<<" herr "<<herr;
                }
            }
        }
    }
    double timeFinsh = (double) clock();
    cout<<"# 3 #"<<"Time = "<<timeFinsh- timeStart<<endl;


    //解锁
    if(0!=UnlockMutexLock(&myMutex3))
    {
        cout<<("Unlock mutex lock fail.\n");
        return nullptr;
    }
    //销毁锁
    if(0!=DestoryMutexLock(&myMutex3))
    {
        cout<<"Destory mutex lock fail.\n";
        return nullptr;
    }
    //线程自身调用pthread_exit()函数，终止线程
//    pthread_exit(0);
}

int main() {

    char errmsg[255];
    int handle_errcode = 0;
    int handle = construct_handle("E:/wizard","refprop.dll", &handle_errcode, errmsg, 255);
    int handle2 = construct_handle("E:/wizard","refprop.dll", &handle_errcode, errmsg, 255);
    int handle3 = construct_handle("E:/wizard","refprop.dll", &handle_errcode, errmsg, 255);

    cout<<"handle="<<handle<<endl;
    cout<<"handle2="<<handle2<<endl;
    cout<<"handle3="<<handle3<<endl;
    cout<<"handle_errcode="<<handle_errcode<<endl;

    int nc = 1;
    int ierr = 0;
    char herr[255], hfld[] = "WATER.FLD", hhmx[] = "HMX.BNC", href[] = "DEF";
    SETUPdll(handle, &handle_errcode, nc, hfld, hhmx, href, ierr, herr, 10000, 255, 3, 255);
    SETUPdll(handle2, &handle_errcode, nc, hfld, hhmx, href, ierr, herr, 10000, 255, 3, 255);
    SETUPdll(handle3, &handle_errcode, nc, hfld, hhmx, href, ierr, herr, 10000, 255, 3, 255);

    TestFunc3(nullptr);

    pthread_t pthread1;
    pthread_t pthread2;
    pthread_create(&pthread1, NULL, TestFunc1, NULL);
    pthread_create(&pthread2, NULL, TestFunc2, NULL);

    pthread_join(pthread1, NULL);
    pthread_join(pthread2, NULL);

    return EXIT_SUCCESS;
}

**Comment by zhaolongjun** [comment]
I tested your example, that ok!
I called  SETUP three times， and no problem with serial operation!

**Comment by zhaolongjun** [comment]
I tested your example, that ok!
I called SETUP three times， and no problem with serial operation , but parallel computing Errors!

**Comment by zhaolongjun** [comment]
I tested your example, that's ok !
I called SETUP three times，got three different handle. As for different handle, serial operation is ok , but parallel computing is easy error!

**Comment by zhaolongjun** [comment]
I used thread locks in the code above. The following didn't use thread lock!

**Comment by zhaolongjun** [comment]
#include "REFPROP-manager/REFPROP-manager.hpp"
using namespace  std;
#include <iostream>
#include<pthread.h>
#include <time.h>
void* thread1(void* arg);
void* thread2(void* arg);

int main()
{
    char errmsg[255];
    int handle_errcode = 0;
    int handle = construct_handle("E:/wizard","refprop.dll", &handle_errcode, errmsg, 255);
    int handle2 = construct_handle("E:/wizard","refprop.dll", &handle_errcode, errmsg, 255);
    int handle3 = construct_handle("E:/wizard","refprop.dll", &handle_errcode, errmsg, 255);
    cout<<"handle="<<handle<<endl;
    cout<<"handle2="<<handle2<<endl;
    cout<<"handle3="<<handle2<<endl;
    cout<<"handle_errcode="<<handle_errcode<<endl;

    int nc = 1;
    int ierr = 0;
    char herr[255], hfld[] = "WATER.FLD", hhmx[] = "HMX.BNC", href[] = "DEF";
    SETUPdll(handle, &handle_errcode, nc, hfld, hhmx, href, ierr, herr, 10000, 255, 3, 255);
    SETUPdll(handle2, &handle_errcode, nc, hfld, hhmx, href, ierr, herr, 10000, 255, 3, 255);
    SETUPdll(handle3, &handle_errcode, nc, hfld, hhmx, href, ierr, herr, 10000, 255, 3, 255);

    int kq=1;
    double z[] = {1.0}, x[] = {1.0}, y[] = {1.0}, T= 500, p = 101.325, d = -1, dl = -1, dv = -1, h = -1, s = -1, u = -1, cp = -1, cv = -1, q = 1, w = -1;
    double timeStart = (double) clock();
    for(int i = 0; i < 1000; i++)
    {
        p +=0.1;
        PQFLSHdll(3,0, p, q, z, kq, T, d, dl, dv, x, y, u, h, s, cp, cv, w, ierr, herr, 255);
        cout<<"# 3 #"<<"p  "<<p<< " h "<<h<<endl;
        if(ierr != 0)
        {
            cout<<"#3 # "<<" ierr "<<ierr<<" herr "<<herr;
        }
    }

    double timeFinsh = (double) clock();
    cout<<"# 3#"<<"Time = "<<timeFinsh- timeStart<<endl;

    pthread_t pthread1;
    pthread_t pthread2;
    pthread_create(&pthread1, NULL, thread1, NULL);
    pthread_create(&pthread2, NULL, thread2, NULL);
    pthread_join(pthread1, NULL);
    pthread_join(pthread2, NULL);

    return EXIT_SUCCESS;
}


void* thread1(void* arg)
{
    int ierr = 0;
    char herr[255];
    int kq = 1;

    double z[] = { 1.0 }, x[] = { 1.0 }, y[] = { 1.0 }, T =500, p = 101.325, d = -1, dl = -1, dv = -1, h = -1, s = -1, u = -1, cp = -1, cv = -1, q = 0, w = -1;
    double timeStart = (double) clock();
    for(int i = 0; i < 1000; i++)
    {
        p += 0.1;
        PQFLSHdll(1,0, p, q, z, kq, T, d, dl, dv, x, y, u, h, s, cp, cv, w, ierr, herr, 255);
        cout<<"# 1 #"<<"p "<<p<< " h "<<h<<endl;
        if(ierr != 0)
        {
            cout<<"# 1 #"<<" ierr "<<ierr<<" herr "<<herr;
        }
    }

    double timeFinsh = (double) clock();
    cout<<"# 1 #"<<"Time = "<<timeFinsh- timeStart<<endl;

}

void* thread2(void* arg)
{

    int ierr = 0;
    char herr[255];
    int kq=1;
    double z[] = {1.0}, x[] = {1.0}, y[] = {1.0}, T=500, p = 101.325, d = -1, dl = -1, dv = -1, h = -1, s = -1, u = -1, cp = -1, cv = -1, q =1, w = -1;
    double timeStart = (double) clock();
    for(int i = 0; i < 1000; i++)
    {
        p +=0.1;
        PQFLSHdll(2,0, p, q, z, kq, T, d, dl, dv, x, y, u, h, s, cp, cv, w, ierr, herr, 255);
        cout<<"# 2 #"<<"P "<<p<< " h "<<h<<endl;
        if(ierr != 0)
        {
            cout<<"# 2 #"<<" ierr "<<ierr<<" herr "<<herr;
        }
    }
    double timeFinsh = (double) clock();
    cout<<"# 2 #"<<"Time = "<<timeFinsh- timeStart<<endl;

}

**Comment by zhaolongjun** [comment]
#include "REFPROP-manager/REFPROP-manager.hpp"
using namespace  std;
#include<pthread.h>
#include <time.h>
#include <iostream>
void* thread1(void* arg);
void* thread2(void* arg);

int main()
{
    char errmsg[255];
    int handle_errcode = 0;
    int handle = construct_handle("E:/wizard","refprop.dll", &handle_errcode, errmsg, 255);
    int handle2 = construct_handle("E:/wizard","refprop.dll", &handle_errcode, errmsg, 255);
    int handle3 = construct_handle("E:/wizard","refprop.dll", &handle_errcode, errmsg, 255);
    cout<<"handle="<<handle<<endl;
    cout<<"handle2="<<handle2<<endl;
    cout<<"handle3="<<handle2<<endl;
    cout<<"handle_errcode="<<handle_errcode<<endl;

    int nc = 1;
    int ierr = 0;
    char herr[255], hfld[] = "WATER.FLD", hhmx[] = "HMX.BNC", href[] = "DEF";
    SETUPdll(handle, &handle_errcode, nc, hfld, hhmx, href, ierr, herr, 10000, 255, 3, 255);
    SETUPdll(handle2, &handle_errcode, nc, hfld, hhmx, href, ierr, herr, 10000, 255, 3, 255);
    SETUPdll(handle3, &handle_errcode, nc, hfld, hhmx, href, ierr, herr, 10000, 255, 3, 255);

    int kq=1;
    double z[] = {1.0}, x[] = {1.0}, y[] = {1.0}, T= 500, p = 101.325, d = -1, dl = -1, dv = -1, h = -1, s = -1, u = -1, cp = -1, cv = -1, q = 1, w = -1;
    double timeStart = (double) clock();
    for(int i = 0; i < 1000; i++)
    {
        p +=0.1;
        PQFLSHdll(3,0, p, q, z, kq, T, d, dl, dv, x, y, u, h, s, cp, cv, w, ierr, herr, 255);
        cout<<"# 3 #"<<"p  "<<p<< " h "<<h<<endl;
        if(ierr != 0)
        {
            cout<<"#3 # "<<" ierr "<<ierr<<" herr "<<herr;
        }
    }

    double timeFinsh = (double) clock();
    cout<<"# 3#"<<"Time = "<<timeFinsh- timeStart<<endl;

    pthread_t pthread1;
    pthread_t pthread2;
    pthread_create(&pthread1, NULL, thread1, NULL);
    pthread_create(&pthread2, NULL, thread2, NULL);
    pthread_join(pthread1, NULL);
    pthread_join(pthread2, NULL);

    return EXIT_SUCCESS;
}


void* thread1(void* arg)
{
    int ierr = 0;
    char herr[255];
    int kq = 1;

    double z[] = { 1.0 }, x[] = { 1.0 }, y[] = { 1.0 }, T =500, p = 101.325, d = -1, dl = -1, dv = -1, h = -1, s = -1, u = -1, cp = -1, cv = -1, q = 0, w = -1;
    double timeStart = (double) clock();
    for(int i = 0; i < 1000; i++)
    {
        p += 0.1;
        PQFLSHdll(1,0, p, q, z, kq, T, d, dl, dv, x, y, u, h, s, cp, cv, w, ierr, herr, 255);
        cout<<"# 1 #"<<"p "<<p<< " h "<<h<<endl;
        if(ierr != 0)
        {
            cout<<"# 1 #"<<" ierr "<<ierr<<" herr "<<herr;
        }
    }

    double timeFinsh = (double) clock();
    cout<<"# 1 #"<<"Time = "<<timeFinsh- timeStart<<endl;

}

void* thread2(void* arg)
{

    int ierr = 0;
    char herr[255];
    int kq=1;
    double z[] = {1.0}, x[] = {1.0}, y[] = {1.0}, T=500, p = 101.325, d = -1, dl = -1, dv = -1, h = -1, s = -1, u = -1, cp = -1, cv = -1, q =1, w = -1;
    double timeStart = (double) clock();
    for(int i = 0; i < 1000; i++)
    {
        p +=0.1;
        PQFLSHdll(2,0, p, q, z, kq, T, d, dl, dv, x, y, u, h, s, cp, cv, w, ierr, herr, 255);
        cout<<"# 2 #"<<"P "<<p<< " h "<<h<<endl;
        if(ierr != 0)
        {
            cout<<"# 2 #"<<" ierr "<<ierr<<" herr "<<herr;
        }
    }
    double timeFinsh = (double) clock();
    cout<<"# 2 #"<<"Time = "<<timeFinsh- timeStart<<endl;

}

**Comment by zhaolongjun** [comment]
#include "REFPROP-manager/REFPROP-manager.hpp"
using namespace  std;
#include<pthread.h>
#include <time.h>
#include <iostream>
void* thread1(void* arg);
void* thread2(void* arg);

int main()
{
    char errmsg[255];
    int handle_errcode = 0;
    int handle = construct_handle("E:/wizard","refprop.dll", &handle_errcode, errmsg, 255);
    int handle2 = construct_handle("E:/wizard","refprop.dll", &handle_errcode, errmsg, 255);
    int handle3 = construct_handle("E:/wizard","refprop.dll", &handle_errcode, errmsg, 255);
    cout<<"handle="<<handle<<endl;
    cout<<"handle2="<<handle2<<endl;
    cout<<"handle3="<<handle2<<endl;
    cout<<"handle_errcode="<<handle_errcode<<endl;

    int nc = 1;
    int ierr = 0;
    char herr[255], hfld[] = "WATER.FLD", hhmx[] = "HMX.BNC", href[] = "DEF";
    SETUPdll(handle, &handle_errcode, nc, hfld, hhmx, href, ierr, herr, 10000, 255, 3, 255);
    SETUPdll(handle2, &handle_errcode, nc, hfld, hhmx, href, ierr, herr, 10000, 255, 3, 255);
    SETUPdll(handle3, &handle_errcode, nc, hfld, hhmx, href, ierr, herr, 10000, 255, 3, 255);

    int kq=1;
    double z[] = {1.0}, x[] = {1.0}, y[] = {1.0}, T= 500, p = 101.325, d = -1, dl = -1, dv = -1, h = -1, s = -1, u = -1, cp = -1, cv = -1, q = 1, w = -1;
    double timeStart = (double) clock();
    for(int i = 0; i < 1000; i++)
    {
        p +=0.1;
        PQFLSHdll(3,0, p, q, z, kq, T, d, dl, dv, x, y, u, h, s, cp, cv, w, ierr, herr, 255);
        cout<<"# 3 #"<<"p  "<<p<< " h "<<h<<endl;
        if(ierr != 0)
        {
            cout<<"#3 # "<<" ierr "<<ierr<<" herr "<<herr;
        }
    }

    double timeFinsh = (double) clock();
    cout<<"# 3#"<<"Time = "<<timeFinsh- timeStart<<endl;

    pthread_t pthread1;
    pthread_t pthread2;
    pthread_create(&pthread1, NULL, thread1, NULL);
    pthread_create(&pthread2, NULL, thread2, NULL);
    pthread_join(pthread1, NULL);
    pthread_join(pthread2, NULL);

    return EXIT_SUCCESS;
}


void* thread1(void* arg)
{
    int ierr = 0;
    char herr[255];
    int kq = 1;

    double z[] = { 1.0 }, x[] = { 1.0 }, y[] = { 1.0 }, T =500, p = 101.325, d = -1, dl = -1, dv = -1, h = -1, s = -1, u = -1, cp = -1, cv = -1, q = 0, w = -1;
    double timeStart = (double) clock();
    for(int i = 0; i < 1000; i++)
    {
        p += 0.1;
        PQFLSHdll(1,0, p, q, z, kq, T, d, dl, dv, x, y, u, h, s, cp, cv, w, ierr, herr, 255);
        cout<<"# 1 #"<<"p "<<p<< " h "<<h<<endl;
        if(ierr != 0)
        {
            cout<<"# 1 #"<<" ierr "<<ierr<<" herr "<<herr;
        }
    }

    double timeFinsh = (double) clock();
    cout<<"# 1 #"<<"Time = "<<timeFinsh- timeStart<<endl;

}

void* thread2(void* arg)
{

    int ierr = 0;
    char herr[255];
    int kq=1;
    double z[] = {1.0}, x[] = {1.0}, y[] = {1.0}, T=500, p = 101.325, d = -1, dl = -1, dv = -1, h = -1, s = -1, u = -1, cp = -1, cv = -1, q =1, w = -1;
    double timeStart = (double) clock();
    for(int i = 0; i < 1000; i++)
    {
        p +=0.1;
        PQFLSHdll(2,0, p, q, z, kq, T, d, dl, dv, x, y, u, h, s, cp, cv, w, ierr, herr, 255);
        cout<<"# 2 #"<<"P "<<p<< " h "<<h<<endl;
        if(ierr != 0)
        {
            cout<<"# 2 #"<<" ierr "<<ierr<<" herr "<<herr;
        }
    }
    double timeFinsh = (double) clock();
    cout<<"# 2 #"<<"Time = "<<timeFinsh- timeStart<<endl;

}

**Comment by zhaolongjun** [comment]
#include "REFPROP-manager/REFPROP-manager.hpp"
using namespace  std;
#include<pthread.h>
#include <time.h>
#include <iostream>
void* thread1(void* arg);
void* thread2(void* arg);

int main()
{
    char errmsg[255];
    int handle_errcode = 0;
    int handle = construct_handle("E:/wizard","refprop.dll", &handle_errcode, errmsg, 255);
    int handle2 = construct_handle("E:/wizard","refprop.dll", &handle_errcode, errmsg, 255);
    int handle3 = construct_handle("E:/wizard","refprop.dll", &handle_errcode, errmsg, 255);
    cout<<"handle="<<handle<<endl;
    cout<<"handle2="<<handle2<<endl;
    cout<<"handle3="<<handle2<<endl;
    cout<<"handle_errcode="<<handle_errcode<<endl;

    int nc = 1;
    int ierr = 0;
    char herr[255], hfld[] = "WATER.FLD", hhmx[] = "HMX.BNC", href[] = "DEF";
    SETUPdll(handle, &handle_errcode, nc, hfld, hhmx, href, ierr, herr, 10000, 255, 3, 255);
    SETUPdll(handle2, &handle_errcode, nc, hfld, hhmx, href, ierr, herr, 10000, 255, 3, 255);
    SETUPdll(handle3, &handle_errcode, nc, hfld, hhmx, href, ierr, herr, 10000, 255, 3, 255);

    int kq=1;
    double z[] = {1.0}, x[] = {1.0}, y[] = {1.0}, T= 500, p = 101.325, d = -1, dl = -1, dv = -1, h = -1, s = -1, u = -1, cp = -1, cv = -1, q = 1, w = -1;
    double timeStart = (double) clock();
    for(int i = 0; i < 1000; i++)
    {
        p +=0.1;
        PQFLSHdll(3,0, p, q, z, kq, T, d, dl, dv, x, y, u, h, s, cp, cv, w, ierr, herr, 255);
        cout<<"# 3 #"<<"p  "<<p<< " h "<<h<<endl;
        if(ierr != 0)
        {
            cout<<"#3 # "<<" ierr "<<ierr<<" herr "<<herr;
        }
    }

    double timeFinsh = (double) clock();
    cout<<"# 3#"<<"Time = "<<timeFinsh- timeStart<<endl;

    pthread_t pthread1;
    pthread_t pthread2;
    pthread_create(&pthread1, NULL, thread1, NULL);
    pthread_create(&pthread2, NULL, thread2, NULL);
    pthread_join(pthread1, NULL);
    pthread_join(pthread2, NULL);

    return EXIT_SUCCESS;
}


void* thread1(void* arg)
{
    int ierr = 0;
    char herr[255];
    int kq = 1;

    double z[] = { 1.0 }, x[] = { 1.0 }, y[] = { 1.0 }, T =500, p = 101.325, d = -1, dl = -1, dv = -1, h = -1, s = -1, u = -1, cp = -1, cv = -1, q = 0, w = -1;
    double timeStart = (double) clock();
    for(int i = 0; i < 1000; i++)
    {
        p += 0.1;
        PQFLSHdll(1,0, p, q, z, kq, T, d, dl, dv, x, y, u, h, s, cp, cv, w, ierr, herr, 255);
        cout<<"# 1 #"<<"p "<<p<< " h "<<h<<endl;
        if(ierr != 0)
        {
            cout<<"# 1 #"<<" ierr "<<ierr<<" herr "<<herr;
        }
    }

    double timeFinsh = (double) clock();
    cout<<"# 1 #"<<"Time = "<<timeFinsh- timeStart<<endl;

}

void* thread2(void* arg)
{

    int ierr = 0;
    char herr[255];
    int kq=1;
    double z[] = {1.0}, x[] = {1.0}, y[] = {1.0}, T=500, p = 101.325, d = -1, dl = -1, dv = -1, h = -1, s = -1, u = -1, cp = -1, cv = -1, q =1, w = -1;
    double timeStart = (double) clock();
    for(int i = 0; i < 1000; i++)
    {
        p +=0.1;
        PQFLSHdll(2,0, p, q, z, kq, T, d, dl, dv, x, y, u, h, s, cp, cv, w, ierr, herr, 255);
        cout<<"# 2 #"<<"P "<<p<< " h "<<h<<endl;
        if(ierr != 0)
        {
            cout<<"# 2 #"<<" ierr "<<ierr<<" herr "<<herr;
        }
    }
    double timeFinsh = (double) clock();
    cout<<"# 2 #"<<"Time = "<<timeFinsh- timeStart<<endl;

}

**Comment by zhaolongjun** [comment]
``` C++
#include "REFPROP-manager/REFPROP-manager.hpp"
using namespace  std;
#include<pthread.h>
#include <time.h>
#include <iostream>

void* thread1(void* arg);
void* thread2(void* arg);

int main()
{
    char errmsg[255];
    int handle_errcode = 0;
    int handle = construct_handle("E:/wizard","refprop.dll", &handle_errcode, errmsg, 255);
    int handle2 = construct_handle("E:/wizard","refprop.dll", &handle_errcode, errmsg, 255);
    int handle3 = construct_handle("E:/wizard","refprop.dll", &handle_errcode, errmsg, 255);
    cout<<"handle="<<handle<<endl;
    cout<<"handle2="<<handle2<<endl;
    cout<<"handle3="<<handle2<<endl;
    cout<<"handle_errcode="<<handle_errcode<<endl;

    int nc = 1;
    int ierr = 0;
    char herr[255], hfld[] = "WATER.FLD", hhmx[] = "HMX.BNC", href[] = "DEF";
    SETUPdll(handle, &handle_errcode, nc, hfld, hhmx, href, ierr, herr, 10000, 255, 3, 255);
    SETUPdll(handle2, &handle_errcode, nc, hfld, hhmx, href, ierr, herr, 10000, 255, 3, 255);
    SETUPdll(handle3, &handle_errcode, nc, hfld, hhmx, href, ierr, herr, 10000, 255, 3, 255);

    int kq=1;
    double z[] = {1.0}, x[] = {1.0}, y[] = {1.0}, T= 500, p = 101.325, d = -1, dl = -1, dv = -1, h = -1, s = -1, u = -1, cp = -1, cv = -1, q = 1, w = -1;
    double timeStart = (double) clock();
    for(int i = 0; i < 1000; i++)
    {
        p +=0.1;
        PQFLSHdll(3,0, p, q, z, kq, T, d, dl, dv, x, y, u, h, s, cp, cv, w, ierr, herr, 255);
        cout<<"# 3 #"<<"p  "<<p<< " h "<<h<<endl;
        if(ierr != 0)
        {
            cout<<"#3 # "<<" ierr "<<ierr<<" herr "<<herr;
        }
    }

    double timeFinsh = (double) clock();
    cout<<"# 3#"<<"Time = "<<timeFinsh- timeStart<<endl;

    pthread_t pthread1;
    pthread_t pthread2;
    pthread_create(&pthread1, NULL, thread1, NULL);
    pthread_create(&pthread2, NULL, thread2, NULL);
    pthread_join(pthread1, NULL);
    pthread_join(pthread2, NULL);

    return EXIT_SUCCESS;
}


void* thread1(void* arg)
{
    int ierr = 0;
    char herr[255];
    int kq = 1;

    double z[] = { 1.0 }, x[] = { 1.0 }, y[] = { 1.0 }, T =500, p = 101.325, d = -1, dl = -1, dv = -1, h = -1, s = -1, u = -1, cp = -1, cv = -1, q = 0, w = -1;
    double timeStart = (double) clock();
    for(int i = 0; i < 1000; i++)
    {
        p += 0.1;
        PQFLSHdll(1,0, p, q, z, kq, T, d, dl, dv, x, y, u, h, s, cp, cv, w, ierr, herr, 255);
        cout<<"# 1 #"<<"p "<<p<< " h "<<h<<endl;
        if(ierr != 0)
        {
            cout<<"# 1 #"<<" ierr "<<ierr<<" herr "<<herr;
        }
    }

    double timeFinsh = (double) clock();
    cout<<"# 1 #"<<"Time = "<<timeFinsh- timeStart<<endl;

}

void* thread2(void* arg)
{

    int ierr = 0;
    char herr[255];
    int kq=1;
    double z[] = {1.0}, x[] = {1.0}, y[] = {1.0}, T=500, p = 101.325, d = -1, dl = -1, dv = -1, h = -1, s = -1, u = -1, cp = -1, cv = -1, q =1, w = -1;
    double timeStart = (double) clock();
    for(int i = 0; i < 1000; i++)
    {
        p +=0.1;
        PQFLSHdll(2,0, p, q, z, kq, T, d, dl, dv, x, y, u, h, s, cp, cv, w, ierr, herr, 255);
        cout<<"# 2 #"<<"P "<<p<< " h "<<h<<endl;
        if(ierr != 0)
        {
            cout<<"# 2 #"<<" ierr "<<ierr<<" herr "<<herr;
        }
    }
    double timeFinsh = (double) clock();
    cout<<"# 2 #"<<"Time = "<<timeFinsh- timeStart<<endl;

}
```

**Comment by ianhbell** [comment]
I'll come up with an example of calling REFPROP in parallel from separate C++11 threads.  

In the meantime, you should fix some things in your code:
* The lengths of x, y, z need to be 20, not 1
* The strings need to be allocated to their full lengths and padded with spaces.

**Comment by zhaolongjun** [comment]
Thanks! I'll do it!

**Comment by zhaolongjun** [comment]
Hello！
Thank you very much for helping to compile the multithreaded examples. Do you have progress on multithreading parallel computation?




 
---Original---

**Comment by ianhbell** [comment]
Have you tried to make the fixes I identified and run your example again?  If you are messing up the memory (as you are if you do not properly allocate memory and REFPROP overwrites random locations in memory), then all bets are off.

**Comment by zhaolongjun** [comment]
If there is no thread locks, mistakes can not be avoided. However, with thread locks, the time cost will be dozens or even hundreds of times that of a single thread. 
I want to know if there are easier ways to implement multithreading, or if different threads call separately and load into different libraries in memory to achieve a one-to-one correspondence between threads and libraries, and each thread does not affect each other.






---Original---

**Comment by ianhbell** [comment]
Please fix the memory allocation problems first before worrying about locks.  I agree that you want to find a solution that doesn’t require locks.

**Comment by zhaolongjun** [comment]
Is the memory allocation error due to multiple calls to SETUPdll?
The handle return value is the number of libraries loaded into memory, right?
How can I load multiple libraries and allocate memory correctly if I don't call SETUPdll many times？



 
---Original---

**Comment by ianhbell** [comment]
> I'll come up with an example of calling REFPROP in parallel from separate C++11 threads.
> 
> In the meantime, you should fix some things in your code:
> 
> * The lengths of x, y, z need to be 20, not 1
> * The strings need to be allocated to their full lengths and padded with spaces.

These memory errors...

**Comment by ianhbell** [comment]
I added an example (test/test2.cpp) showing how to call without any locks with the use of futures.  I have a similar demonstration planned with threadpools.  The current example with futures shows only a modest improvement in speed.  With a threadpool, should be able to achieve nearly linear speedup if the jobs are expensive enough.

**Comment by ianhbell** [comment]
I think this is a fundamental limitation of calling multiple shared objects within the same process.  My suspicion is that multiple calls to dlsym start to collide as you have more threads trying to make the calls, at least that's what it seems like.  I want to see if I can cache these lookups, as it should speed up the resolution, but it will be tricky because each function is of a different type.

**Comment by ianhbell** [comment]
And..  fixed: 60699b69d3195b83d039b9bc9249f2cc18bb0a7f .  In parallel execution, I can yield a speedup now that is close to linear speedup.  Please see test/test2.cpp .  It was indeed a question of caching the pointer to the method.
